cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -940,6 +1319,31 @@ pub enum SchemaTypeVariant<N> {
         /// may not yet be fully qualified
         name: N,
     },
+    /// Reference that may resolve to either an entity or common type
+    EntityOrCommon {
+        /// Name of the entity or common type.
+        /// For the important case of `N` = `RawName`, this is the schema JSON
+        /// format, and the `RawName` is exactly how it appears in the schema;
+        /// may not yet be fully qualified.
+        ///
+        /// There is no possible ambiguity in the JSON syntax between this and
+        /// `Entity`, nor between this and `SchemaType::Common`.
+        /// - To represent a must-be-entity-type reference in the JSON syntax,
+        ///     use `{ "type": "Entity", "name": "foo" }`. This ser/de as
+        ///     `SchemaType::Type(SchemaTypeVariant::Entity)`.
+        /// - To represent a must-be-common-type reference in the JSON syntax,
+        ///     use `{ "type": "foo" }`. This ser/de as
+        ///     `SchemaType::CommonTypeRef`.
+        /// - To represent an either-entity-or-common-type reference in the
+        ///     JSON syntax, use `{ "type": "EntityOrCommon", "name": "foo" }`.
+        ///     This ser/de as `SchemaType::Type(SchemaTypeVariant::EntityOrCommon`.
```

Comment:
Would a user ever want to write this syntax, or is it just used internally / to support round tripping?

Reply:
Based on [this thread](https://github.com/cedar-policy/cedar/pull/1060#discussion_r1692049738), we support users writing this syntax, if they want to.

Reply:
Sure, but there is no reason they would need to, right?

Reply:
Yeah, I don't expect users to write this. Presumably they know which it is, and I'd encourage users to be explicit.

Reply:
There's no reason they would need to, but they might find it convenient, who knows.

I see no reason to try to artificially prevent users from writing this syntax.  It would cause problems for roundtrip properties too.